sync2/multipeer/multipeer.go

```diff
@@ -0,0 +1,463 @@
+				zap.Stringer("peer", p),
+				zap.Float64("sim", pr.Sim),
+				zap.Int("localCount", c))
+			s.nearFullCount++
+		} else {
+			mpr.logger.Debug("nearFull peer",
+				zap.Stringer("peer", p),
+				zap.Float64("sim", pr.Sim),
+				zap.Int("localCount", c))
+		}
+	}
+	return s, nil
+}
+
+func (mpr *MultiPeerReconciler) needSplitSync(s syncability) bool {
+	if float64(s.nearFullCount) >= float64(len(s.syncable))*mpr.minCompleteFraction {
+		// enough peers are close to this one according to minhash score, can do
+		// full sync
+		mpr.logger.Debug("enough peers are close to this one, doing full sync",
+			zap.Int("nearFullCount", s.nearFullCount),
+			zap.Int("peerCount", len(s.syncable)),
+			zap.Float64("minCompleteFraction", mpr.minCompleteFraction))
+		return false
+	}
+
+	if len(s.splitSyncable) < mpr.minSplitSyncPeers {
+		// would be nice to do split sync, but not enough peers for that
+		mpr.logger.Debug("not enough peers for split sync",
+			zap.Int("splitSyncableCount", len(s.splitSyncable)),
+			zap.Int("minSplitSyncPeers", mpr.minSplitSyncPeers))
+		return false
+	}
+
+	mpr.logger.Debug("can do split sync")
+	return true
+}
+
+func (mpr *MultiPeerReconciler) fullSync(ctx context.Context, syncPeers []p2p.Peer) error {
+	var eg errgroup.Group
+	for _, p := range syncPeers {
+		syncer := mpr.syncBase.Derive(p)
+		eg.Go(func() error {
+			defer syncer.Release()
+			err := syncer.Sync(ctx, nil, nil)
+			switch {
+			case err == nil:
+				mpr.sl.NoteSync()
+			case errors.Is(err, context.Canceled):
+				return err
+			default:
+				// failing to sync against a particular peer is not considered
+				// a fatal sync failure, so we just log the error
+				mpr.logger.Error("error syncing peer", zap.Stringer("peer", p), zap.Error(err))
+			}
+			return syncer.Release()
+		})
+	}
+	return eg.Wait()
+}
+
+func (mpr *MultiPeerReconciler) syncOnce(ctx context.Context, lastWasSplit bool) (full bool, err error) {
+	var s syncability
+	for {
+		syncPeers := mpr.peers.SelectBest(mpr.syncPeerCount)
+		mpr.logger.Debug("selected best peers for sync",
+			zap.Int("syncPeerCount", mpr.syncPeerCount),
+			zap.Int("totalPeers", mpr.peers.Total()),
+			zap.Int("numSelected", len(syncPeers)))
+		if len(syncPeers) != 0 {
+			// probePeers doesn't return transient errors, sync must stop if it failed
+			mpr.logger.Debug("probing peers", zap.Int("count", len(syncPeers)))
+			s, err = mpr.probePeers(ctx, syncPeers)
+			if err != nil {
+				return false, err
+			}
+			if len(s.syncable) != 0 {
+				break
+			}
+		}
+
+		mpr.logger.Debug("no peers found, waiting", zap.Duration("duration", mpr.noPeersRecheckInterval))
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		case <-mpr.clock.After(mpr.noPeersRecheckInterval):
+		}
+	}
+
+	full = false
+	if !lastWasSplit && mpr.needSplitSync(s) {
+		mpr.logger.Debug("doing split sync", zap.Int("peerCount", len(s.splitSyncable)))
+		err = mpr.runner.SplitSync(ctx, s.splitSyncable)
```

Comment:
Syncing against the peers that you're in sync with is very cheap as it involves only getting the peer's toplevel fingerprint which is always readily available and is `O(1)` to get for the whole set b/c of how FPTree works (#6406). On the other hand, if some of the peers used for the split sync turn out to have additional items falling into ranges not used during the split, it would be better to get these too in the full sync round

Reply:
The split sync itself usually means that the node is relatively far behind, so it may take some time to complete given the need to validate the ATXs etc., and during cycle gaps this means that each peer involved in split sync may have been getting quite a few new ATXs while the sync was happening. It is worth noting that sync is done using snapshots of the peers' sets taken at the point of its start, so full sync after split sync means doing the necessary catching up